### ISSUE TITLE
Cmd+click to linkify file in terminal doesn't work when there are whitespace or certain separators in the filename

### ISSUE THREAD

**Issue by wzxu** [zed-industries/zed#12338]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

For example, these 3 filenames won't work:
* 'test file 1.txt'
* test、2.txt
* test。3.txt
<img width="520" alt="62246" src="https://github.com/zed-industries/zed/assets/5619348/9f2e819a-9bc7-49a4-81e9-29983e58dbc0">

### Environment

Zed: v0.137.2 (Zed Preview)
OS: macOS 14.4.1
Memory: 16 GiB
Architecture: aarch64

**Comment by SomeoneToIgnore** [comment]
See https://github.com/zed-industries/zed/issues/8256#issuecomment-1960944540 for a similar bug the explanation and https://github.com/zed-industries/zed/pull/8321 and https://github.com/zed-industries/zed/pull/2899 for the PR example.

**Comment by SomeoneToIgnore** [comment]
Not sure the issue could be solved by just accepting more spaces for the string.
We seem to need to start supporting some ways of finding `'`-quoted strings, which seems conditional already, making regex usage harder?

There are few hacks around the `[]` matching already:
https://github.com/zed-industries/zed/blob/dcb4f73540fb8ab4c942a50148e8956282383bc3/crates/terminal/src/terminal.rs#L827
so we might do the following:
* extract the entire "find all path-like things" into a separate method/function, so we can test it somehow + add a few tests

If we start to ingrain more logic, we'd better be sure to test its correctness.

* after tests for all desired cases are written, it's easier to check multiple approaches.
I do not see anything good in particular, but overall it does not seem feasible to include that into `word_regex` — instead, I'd have another regex or logic to match texts between two `'` quotes — note that the quotes can be nested (e.g. due to escaping `'my very weird file with\' symbol inside.bat'`, so not sure that regex is a good solution here.

* On top of that, we might want to deduplicate multiple matches, if one contains another.

**Comment by Spoutnik97** [comment]
Same issue when using a command line tool like jest, tsc or knip

Zed does not detect the path when I cmd+hover the path

**Comment by JAForbes** [comment]
clicking file protocol paths aren't working for me in Zed Dev 0.154.0, e.g. clicking any of these does nothing:

![image](https://github.com/user-attachments/assets/811cedb0-4693-494e-9792-fafb966293c4)

Oddly enough they highlight on hover, so seems they are recognized as links but then nothing happens when you click.

**Comment by davewa** [comment]
@karthiknathan06, I'd like to help with the fix if you are open to that. I created #24244 before finding this issue.

@SomeoneToIgnore, for the requirements/design what about enabling user config of a list of `path_regexes` to allow people to decide for their specific use cases what they would like to support. At least for rustc output cases, the paths are not surrounded in any well-known delimiters and in general I think we should support such cases. My local implementation currently uses regexes to identify lines which are well-known to end in a path, and captures that part of the line in a named capture. 

```
const RUSTC_DIAGNOSTIC_PATH_REGEX: &str = r#"\s+(at|:::|-->) (?<path>.+)$"#;
const RUSTC_PACKAGE_PATH_REGEX: &str = r#"\s+(Compiling|Checking|Documenting) [^\(\n]+\((?<path>[^\)]+)\)"#;

```
This is working well for me for rustc output. Thoughts on adding user config for providing regexes like these? Of course we can still add built-in support for well-known surrounding delimiters.

**Comment by davewa** [comment]
I went ahead and put together a prototype to facilitate design/discussion/testing here: https://github.com/davewa/zed/tree/terminal-hyperlinks.

**Comment by davewa** [comment]
Thanks for looking at my prototype and sending your feedback. I think I agree with all your points, but I'm not sure my prototype is the best approach. I've been thinking about your earlier comment that regex might not be the right solution, and I think you may be right. I have an idea for another approach that I think is better than my prototype in many respects (and which is probably similar to what you were thinking of when you mentioned nested quotes above...). I would like to discuss the design a bit more before I proceed further.

**Design**

The below non-regex approach could be used for linkifying paths with spaces in the terminal without adding any user config, which seems like a big benefit.

```rust
let focused = /* word under the cursor */;
let potential_start_words = &[/* all words on the same line up to focused */];
let potential_end_words = &[/* all words on the same line from focused to the last word */];
let linkified_path = /* longest string starting with a word in potential_start_words and 
    ending with a word in potential_end_words that is a valid path */;
```

**Benefits**
- No configuration needed, it just works for most (all?) cases.
- Handles nested quotes, parenthesis, etc.

**Drawbacks**
- Does a little more work in the default case (bounded by line length though). If this unacceptable, we could retain the existing behavior of checking the focused word first, and not check further if it is a valid path. Also other mitigations against pathological cases (e.g., very long lines) are trivial.
- It is possible to match an existing file for a focused word that is not _meant_ to be a path, e.g. An existing file named "hello to the world" would be linkified in the sentence "I would like to say hello to the world!". I feel that this would be rare and also, _it matches the existing behavior_ ;), ...and is not solvable in a fully general way.

**Considerations**
- Given the potential drawbacks, it seems best to have this disabled by default, so we would need a setting value to explicitly enable it. I think it _would_ be nice to have it enabled by default, if that were deemed acceptable, but I'd rather have it opt-in than not at all.
- The workspace could tell us if any path in the workspace contains spaces, and we could enable this behavior by default if it does. Of course, it could still be disabled in settings.
- It would also be possible to match on strings enclosed in well-known (or user configured) path-surrounding characters; ' " ( )
- It seems desirable to find the longest match, but I could be wrong.

**Comment by SomeoneToIgnore** [comment]
Honestly, not sure I see any benefits over the regexes as, indeed, we seem to replace regex matching with more heuristics (checking for whitespaces in paths and removing the surrounding symbols) with more FS checks required for this.
We're not going away from regexes fully either, as we're forced to check for URLs at least.

Maybe, I'm missing something and it's simple to prototype this to prove me wrong?

**Comment by davewa** [comment]
From an implementation perspective, I see them about the same. Where I see real benefits of the non-regex approach are in usability and functionality. For usability, with the regex approach, everyone who wants this feature enabled has to figure out for their specific language and tools which regexes will work for them, and there are 100s of tools at least. Sure any given developer may only have 10s they are working with, but even that is cumbersome. Also, it is brittle; any change in the tools' output can break everyone's regex. For functionality, the regex approach doesn't work for nested quotes, (or ( ) [ ], etc.) or multiple paths on a line (e.g., the default output of `ls` wouldn't linkify). With the non-regex approach, the only configuration needed for everyone's toolset to just work is `enable_linkify_all_valid_paths = true`, it is not brittle, and it works in cases where regex is difficult or impossible. Seems like a good trade for a few extra `fs.exists()` calls, no? I'm happy to put together a prototype for the non-regex approach--I'll work on that and we can think it over in the meantime.

**Comment by davewa** [comment]
I've pushed the new non-regex prototype here: https://github.com/davewa/zed/tree/terminal-hyperlinks-all-valid-paths
Seems to work nicely. Also, as I was writing that up, I had another idea: _If_ we are willing to limit support to files belonging to the workspace, couldn't we utilize the existing index (⌘-p) for a cleaner solution?

**Comment by SomeoneToIgnore** [comment]
* Zed codebase prefers to avoid using `std::fs` wherever possible and use `trait Fs` instead: https://github.com/zed-industries/zed/blob/6367754737a3fa4b93c32f8ebd9fe044859f441b/crates/fs/src/fs.rs#L79

Note that this is async and won't work well with the event loop most probably.

*
> couldn't we utilize the existing index

That's what we do exactly: https://github.com/zed-industries/zed/blob/6367754737a3fa4b93c32f8ebd9fe044859f441b/crates/terminal_view/src/terminal_view.rs#L851-L866

On top of that, we do check the relative paths for existence in the worktree the same way.
Note that `terminal.rs` is architectured so that it does not know anything about Zed's internals including `Workspace`

Moreover, this workspace index is not complete and will never include `excluded_files` and gitignored files deep in the tree (until their parents are expanded in the project panel or them/neighbors are opened first).

So this is not a solution, but still a good thing to try first before accessing the FS.


* Do I understand right that `LINE_COLUMN_REGEX` assumes `:` as a separator always?
This will hurt MSBuild output format at least: https://github.com/zed-industries/zed/issues/15412 

Also, unfortunately, all `:`, `,`, `(`, `)` symbols are valid in paths of certain OSes:
![Image](https://github.com/user-attachments/assets/0ef04527-64e8-4c45-bc7c-93d1c568aa31)

As nowadays MSBuild is cross-platform, it can generate horrendous edge cases where `New File.txt(2)` could be both an existing file and a link to the 2nd line of another file.

In the new, generic system, how are we going to help both groups that want and do not want MSBuild in their systems?

* Last but not least: traversing lines incrementally, matching them by regexes and doing synchronous FS API calls, totally ignoring existing `MaybeNavigationTarget::PathLike` and relative file resolution, all done in the event loop makes me not enthusiastic at all about the new approach.

**Comment by davewa** [comment]
That's all really helpful feedback, thanks. For some of those points, I think we were already in a agreement--I am planning to do a lot more work on the prototype.

>  So this is not a solution, but still a good thing to try first before accessing the FS.

This is great! I will update the prototype to integrate with the `terminal.rs` architecture and logic in `terminal_view.rs`

> Do I understand right that LINE_COLUMN_REGEX assumes : as a separator always?

It does, and that is a bug...

> In the new, generic system, how are we going to help both groups that want and do not want MSBuild in their systems?

My plan for this is the same as for the case of the path being surrounded by common symbols: use an fs check for the longest potential path, then successively strip off a small, well-defined set of common symbols and possible line/column suffix and do an fs check for each shorter variation.

> Last but not least: traversing lines incrementally, matching them by regexes and doing synchronous FS API calls, totally ignoring existing MaybeNavigationTarget::PathLike and relative file resolution, all done in the event loop makes me not enthusiastic at all about the new approach.

My plan is to update the latest prototype to incorporate all your feedback thus far, including using async Fs, using the index before accessing the FS, integrating with the logic in `terminal_view.rs`, supporting the surrounded by common symbols case, and adding lots of tests.

**Comment by SomeoneToIgnore** [comment]
> and adding lots of tests

Note that this might be hard due to 0 visual tests in Zed and 0 tests related to terminal (modulo regexes but even those are merely "matches or not" checks).

Good luck nonetheless.

**Comment by davewa** [comment]
I've updated https://github.com/davewa/zed/tree/terminal-hyperlinks-all-valid-paths to the point where I'd like additional review and feedback before continuing further. In my testing it is performing and working very well, but much more testing is needed. I do have one big "kitchen sink" test and a few other smaller tests, but there are many edge cases and nuance in this space so I am planning to write many more unit tests for specific cases. All the tests pass on macOS, Ubuntu, and Windows. I'm hopeful that we can reach consensus on the heuristic approach. The prototype currently has both the regex support and the heuristics in a unified design.

**Comment by probably-neb** [comment]
Hey @davewa, I looked over your fork a bit, and I think it's looking pretty good overall. I think the best way to proceed is to create a pull request, mark it as a draft and ping me in the comments. That way we can continue discussing implementation details and see up to date diffs more easily. 

I'd also appreciate if you'd make sure the box is checked that says "allow maintainers to edit the pull request" as that usually helps make things go a bit faster and smoother. Thanks!

**Comment by davewa** [comment]
Sounds good, thanks for looking at it! I'll get a draft PR going next week.

**Comment by davewa** [comment]
Based on the discussions here and on #27218, I propose the following next steps:

#### Add tests for existing functionality

As @SomeoneToIgnore points out near the [beginning of this issue](https://github.com/zed-industries/zed/issues/12338#issuecomment-2155302926), the first PR(s) should be adding tests for existing functionality (no new functionality). I propose the tests would have input of a mock fs, mock worktree(s) (local and remote), terminal content, working directory, and hovered point. The output would include the hyperlinked match, all maybe paths checked, where they were checked (worktree or file), and any open target if one is found. These would be split between terminal and terminal_view crates and would include any refactoring necessary to enable testing. They should pass on macOS, Linux, and Windows. In order to make the tests maintainable, I've been looking into testing crates with support for snapshot testing, and [insta](https://docs.rs/insta/latest/insta/index.html) looks nice. I don't see it being used anywhere in Zed yet though? I feel this will be super helpful in improving maintainability, especially having different snapshot folders for different OSes, and potentially also for different levels of enabled functionality. The snapshot approach should also help keep build times fast as test coverage increases.

#### Design new functionality

There have been some small comments/discussion on various threads--I'd like to write up the design in one place and check it in. My preference for this is a markdown document, with as simple a description of the requirements and design as possible, plus some illustrative examples.

#### Implement new functionality

Once the design is complete, create small, focused PRs for each step of implementation. This may end up being exactly [as proposed by @SomeoneToIgnore](https://github.com/zed-industries/zed/pull/27218#pullrequestreview-2710424964):
- Support for regexes + background fs checks and timeouts
- Support for levels of functionality, e.g., optional extended `/paths with spaces/` support

(And also my proposal):
- Reuse navigation target computed on hover for open

Although after completing the design we may decide on a slightly different set of PRs.

I'll be working on testing existing functionality next (with [insta](https://docs.rs/insta/latest/insta/index.html)).

**Comment by probably-neb** [comment]
Hey @davewa, I appreciate the effort you're putting in to improve this area. 

I have one nit with what you have described here, however, and that is that I don't see a reason to make this more complicated than it needs to be. I think the important part here is just that we identify URLs & paths accurately and _quickly_ (as is reasonable). Once we have the path/URL, it seems trivial to figure out whether it's actually a path in the file system, or send it to the browser if it isn't. Additionally, as far as I can tell, we have had very few issues with the file path pointing to the wrong place (i.e. identifying the wrong file based on the path) so I really don't think testing this step is that important. Therefore, I personally don't see a reason why we would need mock file systems, or snapshot testing, when some simple `text input -> Option<URL>` tests would do just fine.

Beyond just simplicity, however, my understanding is that snapshot testing, by definition, requires some way of "fetching" the previous snapshot. This would be a fairly substantial change to the way our testing and CI works currently, and is almost certainly not something that should be included as part of a small refactor such as this one.


### Suggestions

From my perspective, the reason that we're looking to improve/restructure this codepath, is that the current implementation is complex and hard to modify to extend it's functionality. 

What I'd be interested in is something along the lines of a simple `text in -> Option<URL>`  out structure. Where we have a few builtin parsers (preferably hand written rather than regex, but regex is fine) for the most common cases like `{path}:{row}:{col}`,  `https://{url}`, etc. With the capability for the user to define regex patterns themselves in their settings. 

With that as the base, it becomes much simpler to just extract paths & URLs from terminal text as we come across them, filter the ones that don't exist out (based on the actual file system), link them to a location in the terminal, and reuse that parsed path/url when the user is hovering & subsequently requesting to follow it as you also mentioned. 

With this structure, It also becomes much simpler to test, as we just need a string with a URL or edge case in it to test the URL extraction, and, if we end up wanting tests for the "filter" step, we can deal with MockFS or some other thing that makes sense (we already have ways to create _real_ temporary files/folders in tests) as needed.

With the user-provided regexes, we can also provide some example regexes for styles that are niche enough that they're not worth supporting natively, but worth making available to users in such a way that they just have to paste it in their settings. 

And, if we want to maintain the "parse on demand" behavior we have right now (where urls are only searched for once hovering occurs), that would require a very minimal change to what I've described so far, that being once the user hovers, we extract some bounded region around the cursor and try to parse.

But that's just my two cents, and is in not gospel in any way so feel free to disagree. I'm happy to create a rough first draft of the first steps (extracting out the current implementation to work on text rather than terminal grid) if my explanation isn't making sense for you.

**Comment by davewa** [comment]
So far I've only refactored FindHyperlink processing in terminal.rs (just to improve it's testability). I haven't done anything with mocks or insta yet. I share your desire for simplicity, and excluding paths with spaces, I think something like what you suggest would be nice. How about this: I add some simple tests for existing functionality to my refactoring of FindHyperlink, and share those before doing anything else? I can post another draft PR if that is the preferred workflow. I think the test coverage is the most important thing that is lacking currently, and looking at actual tests is a great way to build a common understanding of what the use cases are and how much complexity is actually necessary to support those. Currently on stable for example, any path with a unicode character in it is failing to linkify for me--that seems like something that really should never be broken. As far as testing just text input, I'm worried that would miss coverage of scenarios that are cell-specific, for example with single unicode characters that are two cells wide (the char in the first cell and wide char spacer in the second cell). How would we write a test which asserts that the link underlines both cells? I'm just learning about how this is handled in alacritty, so my understanding of this is probably not complete, but it does seem like something that should have test coverage.

**Comment by davewa** [comment]
I should note that my original regex approach does of course work for _some specific_ paths with spaces cases, and we could add (or allow users to add) a regex for each case, i.e., cargo output, rustc diagnostics, python diagnostics, various forms of ls/dir output, etc.. This could end up being a long list of regexes for some people. The heuristic approach in #27218 is required for 1) enabling entire classes of these types of use cases _by default_ and also 2) some use cases which would be extremely difficult (impossible?) with regex. I totally understand if we're not interested in 1) or 2), and I'd be happy to proceed with the regex approach first, get that merged, and then contemplate the demand/cost/benefit of doing further work on the heuristic approach after that.

**Comment by davewa** [comment]
Perhaps it would be helpful to summarize where I think we are with this? I clearly went a little overboard by fixing a lot of separate issues at the same time in #27218:

1. (this issue) Paths with unicode characters that don't match `\w`
2. (this issue) Paths with spaces
3. Adding a setting for user regexes for identifying paths and optional positions
5. Ambiguous paths, i.e., `~`,`/a`, `/b`, `:4:2`, `(4,2)`, `'hello.rs'`, etc., are all valid real paths, but we parse them as special cases
6. Reusing open target computed for hover when navigating
7. Supporting all valid paths for each OS, e.g., many valid Linux paths are invalid Windows paths

I feel that each of these should separate issues (even though some do overlap a bit), each with their own cost/benefits/tradeoffs which we may or may not deem worth fixing. I think it would be nice to start with the simplest which have the most consensus (1, 3, and 6?) and then look at the others.

**Comment by probably-neb** [comment]
> for example with single unicode characters that are two cells wide (the char in the first cell and wide char spacer in the second cell). How would we write a test which asserts that the link underlines both cells?

I think this is the kind of thing we should fix when it becomes a problem. My intuition is that it's very rare for this to even occur, and not worth structuring our  solution / testing around (not saying that you are, just to illustrate the point). Worth keeping in mind for sure, so we don't back ourselves into a corner, but not something to super worry about is my takeaway.

Thanks for the summary @davewa, very helpful! I also completely agree. I think 1,3, and 5 are a great place to start. I also took a look at your branch and I think it's a great first step. Feel free to open a PR for it once you're ready. We should also probably create a discussion for this to avoid polluting / pinging people on this this issue.

**Comment by probably-neb** [comment]
Ok, created a discussion: [Terminal Hyperlink Detection Refactor · zed-industries/zed · Discussion #28238](https://github.com/zed-industries/zed/discussions/28238)

**Comment by davewa** [comment]
I don't think #37684 addresses this issue (I should pay closer attention to the _description_ when reviewing...)

**Comment by davewa** [comment]
Can this be reopened?
